contracts/rbn-staking/PenaltyDistributor.vy

```diff
@@ -418,11 +433,31 @@ def donate(_amount: uint256) -> bool:
 
     ERC20(self.token).transferFrom(msg.sender, self, _amount)
 
+    penalty_rebate: uint256 = self.penalty_rebate_of[tx.origin]
+    if(block.timestamp < self.penalty_rebate_expiry and penalty_rebate > 0):
+        # Penalty rebate is how much we need to give back to the user
+        # To achieve 50% penalty-free unlock, the original penalty (_amount)
+        # must be divided by two and returned back to the user. This is equivalent
+        # to 50% penalty-free unlock + 50% existing mechanism penalty
+
+        # min = penalty_rebate if user locked in more since RGP-31 snapshot
+        # min = _amount / 2 if user has performed no actions since RGP-31 (increase lock amount or time)
+        ERC20(self.token).transfer(tx.origin, min(penalty_rebate, _amount / 2))
+        self.penalty_rebate_of[tx.origin] = 0
```

Comment:
Is there a possibility that tx.origin is not on the approx 800 addresses of the csv file and the rebate that is being set to 0 is not from that list?

Reply:
people who should not be getting a rebate should still be able to force_withdraw (just w/o rebate)

Reply:
An address only passes this check `if(block.timestamp < self.penalty_rebate_expiry and penalty_rebate > 0)` if it's on the csv list - penalty_rebate > 0 - so in a way tx.origin is limited to addresses on the list - which is good